packages/core/src/types/json-schema-adapter.ts

```diff
@@ -0,0 +1,217 @@
+          );
+        }
+        const discriminatedProperty = option?.properties?.[discriminatedPropertyName];
+        if (!discriminatedProperty || typeof discriminatedProperty === 'boolean') {
+          throw new Error(
+            `Unexpected property '${discriminatedProperty}' schema in ${propertyName} at position ${i} in oneOf.`
+          );
+        }
+        const name = discriminatedProperty.const as string;
+        return transformJSONSchemaToNodeType(name, option, ctx);
+      });
+
+      return (value: unknown, key: string) => {
+        if (isPlainObject(value)) {
+          const discriminatedTypeName = value[discriminatedPropertyName];
+          if (typeof discriminatedTypeName === 'string' && ctx[discriminatedTypeName]) {
+            return discriminatedTypeName;
+          }
+        }
+        return findOneOf(schema.oneOf as JSONSchema[], oneOfs)(value, key);
+      };
+    } else {
+      const oneOfs = schema.oneOf.map((option, i) =>
+        transformJSONSchemaToNodeType(propertyName + '_' + i, option, ctx)
+      );
+      return findOneOf(schema.oneOf as JSONSchema[], oneOfs);
+    }
+  }
+
+  return schema as PropType;
+}
+
+function extractNodeToContext(
+  propertyName: string,
+  schema: JSONSchema,
+  ctx: Record<string, NodeType>
+): string {
+  if (!schema || typeof schema === 'boolean') {
+    throw new Error(`Unexpected schema in ${propertyName}.`);
+  }
+
+  if (schema instanceof Array) {
+    throw new Error(`Unexpected array schema in ${propertyName}. Try using oneOf instead.`);
+  }
+
+  if (schema.type === 'null') {
+    throw new Error(`Unexpected null schema type in ${propertyName} schema.`);
+  }
+
+  if (schema.type instanceof Array) {
+    throw new Error(
+      `Unexpected array schema type in ${propertyName} schema. Try using oneOf instead.`
+    );
+  }
+
+  const properties: Record<string, PropType | ResolveTypeFn> = {};
+  for (const [name, property] of Object.entries(schema.properties || {})) {
+    properties[name] = transformJSONSchemaToNodeType(propertyName + '.' + name, property, ctx);
+  }
+
+  let additionalProperties;
+  if (isPlainObject(schema.additionalProperties)) {
+    additionalProperties = transformJSONSchemaToNodeType(
+      propertyName + '_additionalProperties',
+      schema.additionalProperties,
+      ctx
+    );
+  }
+  if (schema.additionalProperties === true) {
+    additionalProperties = {};
+  }
+
+  let items;
+  if (
+    isPlainObject(schema.items) &&
+    (isPlainObject(schema.items.properties) ||
+      isPlainObject(schema.items.additionalProperties) ||
+      schema.items.oneOf) // exclude scalar array types
+  ) {
+    items = transformJSONSchemaToNodeType(propertyName + '_items', schema.items, ctx);
+  }
+
+  let required = schema.required as NodeType['required'];
+  // Translate required in oneOfs into a ResolveTypeFn.
+  if (schema.oneOf && schema.oneOf.every((option) => !!(option as Oas3Schema).required)) {
```

Comment:
`option.required` can be an empty array?

Reply:
technically yes